topostats/statistics.py

```diff
@@ -0,0 +1,81 @@
+
+import numpy as np
+import pandas as pd
+
+from topostats.logs.logs import setup_logger, LOGGER_NAME
+from topostats.roughness import roughness_rms
+
+LOGGER = setup_logger(LOGGER_NAME)
+
+
+def image_statistics(
+    image: np.ndarray,
+    filename: str,
+    pixel_to_nm_scaling: float,
+    results_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """Calculate statistics pertaining to the whole image, for example the size of the image in pixels and
+    metres, the root-mean-squared roughness and the grains per metre squared.
+
+    Parameters
+    ----------
+    image: np.ndarray
+        Numpy 2D image array of the image to calculate stats for.
+    filename: str
+        The name of the file being processed.
+    pixel_to_nm_scaling: float
+        Float of the scaling factor between pixels and nanometres.
+    results_df: pd.DataFrame
+        Pandas DataFrame of statistics pertaining to individual grains including from grainstats and
+        dna tracing.
+
+    Returns
+    -------
+    dict
+        Dictionary of image statistics.
+    """
+
+    image_stats = {
+        "Image": filename,
+        "image_size_x_m": None,
+        "image_size_y_m": None,
+        "image_area_m2": None,
+        "image_size_x_px": image.shape[1],
+        "image_size_y_px": image.shape[0],
+        "image_area_px2": None,
+        "grains_number_above": None,
+        "grains_per_m2_above": None,
+        "grains_number_below": None,
+        "grains_per_m2_below": None,
+        "rms_roughness": None,
+    }
+
+    # Calculate dimensions of the image
+    image_stats["image_size_x_m"] = image.shape[1] * pixel_to_nm_scaling * 1e-9
+    image_stats["image_size_y_m"] = image.shape[0] * pixel_to_nm_scaling * 1e-9
+    image_stats["image_area_m2"] = image_stats["image_size_x_m"] * image_stats["image_size_y_m"]
+    image_stats["image_area_px2"] = image_stats["image_size_x_px"] * image_stats["image_size_y_px"]
+
+    # Calculate the RMS roughness of the sample on the flattened image.
+    image_stats["rms_roughness"] = roughness_rms(image=image) * 1e-9
+
+    # Calculate image stats relating to grain statistics. Note that the existence of any of these stats
+    # is not guaranteed
+    try:
+        image_stats["grains_number_below"] = results_df["threshold"].value_counts().get("below", 0)
+        image_stats["grains_per_m2_below"] = image_stats["grains_number_below"] / image_stats["image_area_m2"]
+    except KeyError:
+        pass
+    try:
+        image_stats["grains_number_above"] = results_df["threshold"].value_counts().get("above", 0)
+        image_stats["grains_per_m2_above"] = image_stats["grains_number_above"] / image_stats["image_area_m2"]
+    except KeyError:
+        pass
+
+    # Convert the dictionary to a dataframe, with "image_name" as the index. Transpose is needed
+    # else the rows and columns will be swapped.
+    image_stats_df = pd.DataFrame.from_dict(image_stats, orient="index").transpose()
```

Comment:
I think if you use `orient="columns"` then you shouldn't need to `.transpose()`.

Reply:
I had tried that but it produces an error: 

```
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/sylvi/Documents/TopoStats/topostats/processing.py", line 330, in process_scan
    image_stats = image_statistics(
  File "/Users/sylvi/Documents/TopoStats/topostats/statistics.py", line 79, in image_statistics
    image_stats_df = pd.DataFrame.from_dict(image_stats, orient="columns")
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/lib/python3.10/site-packages/pandas/core/frame.py", line 1764, in from_dict
    return cls(data, index=index, columns=columns, dtype=dtype)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/lib/python3.10/site-packages/pandas/core/frame.py", line 664, in __init__
    mgr = dict_to_mgr(data, index, columns, dtype=dtype, copy=copy, typ=manager)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 493, in dict_to_mgr
    return arrays_to_mgr(arrays, columns, index, dtype=dtype, typ=typ, consolidate=copy)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 118, in arrays_to_mgr
    index = _extract_index(arrays)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 656, in _extract_index
    raise ValueError("If using all scalar values, you must pass an index")
ValueError: If using all scalar values, you must pass an index
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/bin/run_topostats", line 8, in <module>
    sys.exit(main())
  File "/Users/sylvi/Documents/TopoStats/topostats/run_topostats.py", line 236, in main
    for img, result, individual_image_stats_df in pool.imap_unordered(
  File "/opt/homebrew/Caskroom/miniforge/base/envs/new/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
ValueError: If using all scalar values, you must pass an index
```

I'm by no means experienced with Pandas and so I could be missing something here. I found the transpose method via a tutorial but I can't remember where

Reply:
I feel bad for suggesting `pd.DataFrame.from_dict()` now, plain `pd.DataFrame()` should do the trick...

```python
import pandas as pd

test_dict = {"a": 1, "b": 2, "c": 3}
pd.DataFrame.from_dict(test_dict, orient="index").T
Out: 
   a  b  c
0  1  2  3
pd.DataFrame([test_dict])
Out: 
   a  b  c
0  1  2  3
```

Sorry :blush: